R/stochastic_process.R

```diff
@@ -76,84 +74,428 @@ stone_stochastic_process <- function(con, modelling_group, disease,
+    index_to <- read_params$index_end
   }
 
-  if (is.data.frame(dalys)) {
-    stopifnot(all.equal(sort(names(dalys)),
-      c("average_duration", "disability_weight", "outcome", "proportion")))
+  if (is.na(index_from)) {
+    index_from <- 1
+    index_to <- 1
   }
 
-  # Scenario-specific tests:
-  # 1. (touchstone, scenario_description) exists in scenario table
-  # 2. (scenario_description, disease) exists in scenario_description table
-  # 3. (touchstone, modelling_group) exists in responsibility_set table
-  # 4. (responsibility_set, scenario) exist in responsibility, where
-  #    scenario id comes from #1, and responsibility_set id from #3
+  ################################################################
 
-  for (scenario in scenarios) {
-    scenario_id <- DBI::dbGetQuery(con, "
-      SELECT id FROM scenario WHERE touchstone = $1
-         AND scenario_description = $2", list(touchstone, scenario))$id
-    if (length(scenario_id) != 1) {
-      stop(sprintf("scenario %s not found in touchstone %s",
-                   scenario, touchstone))
+  for (i in index_from:index_to) {
+
+    the_file <- read_params$files[scenario_no]
+    the_file <- gsub(":index", i, the_file)
+    the_file <- gsub(":group", touchpoint$modelling_group, the_file)
+    the_file <- gsub(":touchstone", touchpoint$touchstone, the_file)
+    the_file <- gsub(":disease", touchpoint$disease, the_file)
+    the_file <- gsub(":scenario", scenario, the_file)
+
+    if (!file.exists(the_file)) {
+      stop(sprintf("File not found: %s", the_file))
     }
+    message(the_file)
 
-    scenario_descs <- DBI::dbGetQuery(con, "
-      SELECT count(*) FROM scenario_description WHERE id = $1
-         AND disease = $2", list(scenario, disease))$count
-    if (scenario_descs != 1) {
-      stop(sprintf("scenario_description %s not valid for disease %s",
-                   scenario, disease))
+    scenario_data[[i]] <-
+      read_xz_csv(con, the_file, outcomes,
+                  read_params$allow_missing_disease,
+                  read_params$runid_from_file, i,
+                  touchpoint$touchstone, countries)
+  }
+
+  # We now have a full scenario. Eliminate age, splitting into
+  # the four files (calendar/cohort, and u5/all age).
+  # For now, in the cohort files, I'm going to call cohort
+  # 'year' - just to keep code tidier, as the code is common...
+
+  if (index_from == index_to) {
+    scenario_data <- scenario_data[[1]]
+  } else {
+    scenario_data <- rbindlist(scenario_data)
+  }
+
+  #######################################################
+
+  agg_and_sort <- function(data) {
+    # Next lines are just to avoid travis NOTEs on
+    # the by = list line.
+    run_id <- NULL
+    year <- NULL
+    country <- NULL
+    data <- data[ , lapply(.SD, sum),
+                  by = list(run_id, year, country),
+                  .SDcols = c("cases", "dalys", "deaths")]
+    data[order(data$run_id, data$country, data$year), ]
```

Comment:
This looks a bit grizzly now. (I didn't know how to use dplyr when I wrote this).

I think this is along the lines of
```
data <- data %>% group_by(run_id, year, country) %>% 
       summarise(cases = sum(cases), dalys = sum(dalys), deaths = sum(deaths))
```

Reply:
Then the comment and those three NULLs can go, and probably there's a %>% sort on dplyr is there?

Reply:
Updated that to use dplyr, will still need the NULL bit to avoid R CMD notes which is a bit annoying